src/UkModulusCheck/Implementation.fs

```diff
@@ -0,0 +1,232 @@
+            let method = xs.[2] |> parseMethod
+            let weightings = xs.[3..16] |> Array.toList |> List.map str2int |> List.sequence
+            let ex = str2int xs.[17]
+            Option.map3 (fun m ws e ->
+                { SortCodeFrom = scFrom; SortCodeTo = scTo; Method = m; Weightings = ws; Exception = Some (Exception e) })
+                method weightings ex
+        | xs when xs.Length = 17 ->
+            let scFrom, scTo = xs.[0], xs.[1]
+            let method = xs.[2] |> parseMethod
+            let weightings = xs.[3..16] |> Array.toList |> List.map str2int |> List.sequence
+            Option.map2 (fun m ws ->
+                { SortCodeFrom = scFrom; SortCodeTo = scTo; Method = m; Weightings = ws; Exception = None })
+                method weightings
+        | _ -> None
+
+    let parseSubstitution (line: string) =
+        match line.Split([|' '|], StringSplitOptions.RemoveEmptyEntries) with
+        | [|scFrom; scTo|] -> Some { SortCode = scFrom; SubstituteWith = scTo }
+        | _ -> None
+
+    let standardise (sortCode: string) (accountNo: string) =
+        match sortCode.Length, accountNo.Length with
+        | 6, 8 -> Some (sortCode, accountNo)
+        | 6, 6 -> Some (sortCode, "00" + accountNo)
+        | 6, 7 -> Some (sortCode, "0" + accountNo)
+        | 6, 9 -> Some (sortCode.[0..4] + accountNo.[0..0], accountNo.[1..]) // For Santander
+        | 6, 10 ->
+            if sortCode.StartsWith("08") // Co-Operative Bank plc
+            then Some (sortCode, accountNo.[..7])
+            else Some (sortCode, accountNo.[2..])
+        | _ -> None
+
+    let doubleAlternateModulus (weightings: Weight list) (number: string) : int =
+        number
+        |> Seq.map char2int
+        |> Seq.map2 (*) weightings
+        |> Seq.collect (Seq.unfold (fun x -> if x > 0 then Some (x % 10, x / 10) else None))
+        |> Seq.sum
+
+    let standardModulus (weightings: Weight list) (number: string) : int =
+        number
+        |> Seq.map char2int
+        |> Seq.map2 (*) weightings
+        |> Seq.sum
+
+    let validateRule rule number =
+        match rule.Method with
+        | Standard10 -> standardModulus rule.Weightings number % 10 = 0
+        | Standard11 -> standardModulus rule.Weightings number % 11 = 0
+        | DoubleAlternate -> doubleAlternateModulus rule.Weightings number % 10 = 0
+
+    let validateRules (rules: ValidationRule list) (substitutionTable: SortCodeSubstitution list) (sortCode: SortCode) (accountNo: AccountNumber) =
+        let number = sortCode + accountNo
+        let exceptions = rules |> List.map (fun r -> r.Exception)
+        let isValid =
+            match rules, exceptions with
+            | [], _ -> true
+
+            | [rule], [None] ->
+                validateRule rule number
+
+            | [rule1; rule2], [None; None] ->
+                (validateRule rule1 number) && (validateRule rule2 number)
+
+            | [rule], [Some ex] ->
+                match ex with
```

Comment:
This internal pattern matching can be flatted to `| [rule], [Some (Exception 1)]` etc, I think I having it on the top level seems more explicit.

Reply:
I've made them nested for better organisation - so it's easier to navigate between the cases, but if you think it's better to flatten it then I can change it.